calicoctl/commands/migrateipam.go

```diff
@@ -0,0 +1,301 @@
+	"github.com/projectcalico/libcalico-go/lib/backend/model"
+	client "github.com/projectcalico/libcalico-go/lib/clientv3"
+	"github.com/projectcalico/libcalico-go/lib/errors"
+	"github.com/projectcalico/libcalico-go/lib/net"
+)
+
+type migrateIPAM struct {
+	client          bapi.Client
+	BlockAffinities []*BlockAffinityKVPair `json:"block_affinities,omitempty"`
+	IPAMBlocks      []*IPAMBlockKVPair     `json:"blocks,omitempty"`
+	IPAMHandles     []*IPAMHandleKVPair    `json:"handles,omitempty"`
+	IPAMConfig      *IPAMConfigKVPair      `json:"config,omitempty"`
+}
+
+type BlockAffinityKVPair struct {
+	Key      *BlockAffinityKey
+	Value    *model.BlockAffinity
+	Revision string
+	TTL      time.Duration // For writes, if non-zero, key has a TTL.
+}
+
+type BlockAffinityKey struct {
+	CIDR net.IPNet `json:"cidr,omitempty"`
+	Host string    `json:"host,omitempty"`
+}
+
+type IPAMBlockKVPair struct {
+	Key      *BlockKey
+	Value    *model.AllocationBlock
+	Revision string
+	TTL      time.Duration // For writes, if non-zero, key has a TTL.
+}
+
+type BlockKey struct {
+	CIDR net.IPNet `json:"cidr,omitempty"`
+}
+
+type IPAMHandleKVPair struct {
+	Key      *model.IPAMHandleKey
+	Value    *model.IPAMHandle
+	Revision string
+	TTL      time.Duration // For writes, if non-zero, key has a TTL.
+}
+
+type IPAMConfigKVPair struct {
+	Key      *model.IPAMConfigKey
+	Value    *model.IPAMConfig
+	Revision string
+	TTL      time.Duration // For writes, if non-zero, key has a TTL.
+}
+
+// ipamResults contains the results from executing an IPAM backend command
+// TODO: May not need this if we decide not to print out the resources that were imported
+type ipamResults struct {
+	// The number of resources that are being configured.
+	numResources int
+
+	// The number of resources that were actually configured.  This will
+	// never be 0 without an associated error.
+	numHandled int
+
+	// The results returned from each invocation
+	resources []*model.KVPair
+
+	// Errors associated with individual resources
+	resErrs []error
+}
+
+func NewMigrateIPAM(c client.Interface) *migrateIPAM {
+	type accessor interface {
+		Backend() bapi.Client
+	}
+	bc := c.(accessor).Backend()
+	return &migrateIPAM{
+		client: bc,
+	}
+}
+
+func (m *migrateIPAM) PullFromDatastore() error {
+	ctx := context.Background()
+
+	blockKVList, err := m.client.List(ctx, model.BlockListOptions{}, "")
+	if err != nil {
+		return err
+	}
+
+	blockAffinityKVList, err := m.client.List(ctx, model.BlockAffinityListOptions{}, "")
+	if err != nil {
+		return err
+	}
+
+	ipamHandleKVList, err := m.client.List(ctx, model.IPAMHandleListOptions{}, "")
+	if err != nil {
+		return err
+	}
+
+	ipamConfigKV, err := m.client.Get(ctx, model.IPAMConfigKey{}, "")
+	if err != nil {
+		// If the resource does not exist, do not throw the error
+		if _, ok := err.(errors.ErrorResourceDoesNotExist); !ok {
+			return err
+		}
+	}
+
+	// Convert all of the abstract KV Pairs into the appropriate types.
```

Comment:
I don't understand why we need to convert these from the abstract kvpairs to typed variants. Could you explain that to me?

Reply:
It is less for going from the golang representation to JSON than it is for going from JSON back to the golang representation. The KVPairs need to have the keys and values as concrete types for going from JSON to the golang objects since otherwise go does not know what object it is actually deserializing into. I could have created a separate `migrateIPAM` object that has the concrete types and used that one solely for going from JSON -> golang and then erase this logic and use this `migrateIPAM` object solely for going from golang -> JSON but I decided that having one object with a little more logic around it looked better to me.